torchtune/utils/_profiler.py

```diff
@@ -4,47 +4,350 @@
+        return_cfg (bool): Return updated profiler config. Default is False.
 
     Returns:
-        ContextManager: PyTorch profiler as a context manager
+        tuple: [torch.profiler.profile, DictConfig]
+
+    NOTE:
+    - Enabling the profiler will result in training speed reduction.
+    - Setting `profile_memory: true` will generate large trace files.
+    - The profiler schedule is context dependent:
+        - Calling `profiler.step()` at each batch iteration but outside the gradient accumulation
+        scope will `step` the profiler each forward / backward step
+        - Calling `profiler.step()` each batch iteration but within the gradient accumulation scope
+        will `step` the profiler each optimizer update step such that each `step` contains multiple
+        forward / backward passes.
+
+    Additional notes:
+        - the profiler schedule updates with respect to an optimizer step:
+            - e.g., if `gradient_accumulation = 2`, then the profiler will step every 2 batches.
+        - sensible defaults will be chosen if the config is missing options
+            - if no activities are specified, profiler will default to CPU + CUDA
+            - if no schedule is specified, profiler will default to wait 10, warmup 5, active 3, repeat 1
+            - if a schedule is specified, profiler will validate that the schedule is valid and can be passed to `instantiate`
+            - certain options will be overridden (`with_stack` and `record_shapes`) depending on requirements of other options
+                - e.g., `profile_memory` requires `with_stack` and `record_shapes`
     """
 
-    def trace_handler(prof) -> None:
-        prof.export_chrome_trace(output_dir)
-
-    return (
-        profile(
-            activities=[
-                torch.profiler.ProfilerActivity.CPU,
-                torch.profiler.ProfilerActivity.CUDA,
-            ],
-            schedule=torch.profiler.schedule(wait=100, warmup=5, active=5, repeat=1),
-            on_trace_ready=trace_handler,
-            record_shapes=True,
-            profile_memory=False,
-            with_stack=False,
+    # Set up profiler activities
+    activities = []
+    if cpu:
+        activities.append(torch.profiler.ProfilerActivity.CPU)
+    if cuda:
+        activities.append(torch.profiler.ProfilerActivity.CUDA)
+    if len(activities) == 0:
+        _warn("No activities specified, defaulting to CPU + CUDA")
+        activities = _DEFAULT_PROFILER_ACTIVITIES
+        cpu = cuda = True
+
+    # Set up profiler schedule
+    use_default_schedule = not any([wait, warmup, active, repeat])
+
+    # Use default schedule if None, else validate that schedule is valid and can be passed to `instantiate`
+    if use_default_schedule:
+        wait = _DEFAULT_SCHEDULE["wait"]
+        warmup = _DEFAULT_SCHEDULE["warmup"]
+        active = _DEFAULT_SCHEDULE["active"]
+        repeat = _DEFAULT_SCHEDULE["repeat"]
```

Comment:
Any reason we handle the defaults for schedule differently than for other configurable fields? We could just fold the values from `_DEFAULT_SCHEDULE` into the corresponding default values in `setup_torch_profiler`. That way it'd be consistent with how we handle the other fields

Reply:
fixed